File: class2relational/src/main/java/org/atlanmod/scratch/dynamic/DynamicElement.java
package org.atlanmod.scratch.dynamic;

import java.util.HashMap;
import java.util.Map;

public class DynamicElement {

    private String type;
    // private String namespace;
    private Map<String, Object> properties;

    public DynamicElement(String type){
        this.type = type;
        properties = new HashMap();
    }

    public DynamicElement(DynamicElement element){
        this.type = element.getType();
        this.properties = element.getProperties();
    }

    public DynamicElement(DynamicElement element, Map<String, Object> properties){
        this.type = element.getType();
        this.properties = properties;
    }

    public  Map<String, Object> getProperties() {
        return this.properties;
    }

    public Object eGet(String name){
        return properties.get(name);
    }

    public String getType(){
        return this.type;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof DynamicElement){
            DynamicElement obj_ = (DynamicElement) obj;
            return this.getType() == obj_.getType() && this.getProperties() == obj_.getProperties();
        }
        return false;
    }

}

File: sparktl-core/src/main/scala/org/atlanmod/tl/model/impl/dynamic/DynamicElement.scala
package org.atlanmod.tl.model.impl.dynamic

import org.atlanmod.tl.util.ListUtils

import scala.collection.mutable

abstract class DynamicElement(classname: String,
                     properties : scala.collection.mutable.Map[String, Any] = mutable.HashMap())
  extends Serializable with ListUtils.Weakable {

//    def this(element: DynamicElement, properties: scala.collection.mutable.Map[String, Any] = null) {
//        this(element.getType, if (properties == null) element.getProperties else properties)
//    }

    def getType: String = { classname }

    def getProperties: scala.collection.mutable.Map[String, Any] = { this.properties }

    def eGetProperty(name: String): Any = {
        this.properties.get(name) match {
            case Some(v) => v
            case _ => null
        }
    }

    def eSetProperty(name: String, value: Any) : Unit = {
        if (!properties.isDefinedAt(name)) properties.put(name, value)
    }

    override def equals(o : Any): Boolean = {
        o match {
            case obj: DynamicElement =>
                this.classname.equals(obj.getType) &&
                  this.properties.equals(obj.getProperties)
            case _ => false
        }
    }

    override def weak_equals(o: Any): Boolean

    override def toString: String = {
        var res = ""
        res += classname + "\n"
        if (properties.nonEmpty) { properties.mkString("\n") }
        res
    }

}
